Add load image support in pointillism [GCI]

srevinsaju opened the pull request:
Fixes https://github.com/sugarlabs/pointillism-activity/issues/17

Pointillism shows a black screen which means nothing for a user who doesn't have a camera
To make it more meaningful, the GCI task requested us to show a message that a camera 
is necessary for pointillism to work.

# [FEATURE PROPOSAL] Load an Image to pointillism
## Issues Faced Previously
pointillism has an issue of not working on systems without a camera.
This PR gives a warning message that a camera is not present if 
`/dev/video.` does not exist.

## Methodology
* Utilize the absence of camera to load an image

## Screenshots
![Screenshot from 2019-12-17 15-48-02](https://user-images.githubusercontent.com/48695438/70996858-bb95a400-20e4-11ea-99f8-22bd0eebc295.png)
Image of Pointiliism showing a friendly Error Message in the absence of camera
Notice the new open toolbar button.
![Screenshot from 2019-12-17 15-53-22](https://user-images.githubusercontent.com/48695438/70997150-6b6b1180-20e5-11ea-96c9-e6154361eb60.png)
An image of the Image Loader from Journal
![Pointillism](https://user-images.githubusercontent.com/48695438/70996965-fd264f00-20e4-11ea-8555-c8677a90a5b0.png)
Image of Pointillism converting a screenshot of speak activity

quozl commented:
Thanks.  Not tested.  Briefly reviewed.  User experience looks about right.  Suggestions;

* [ ] the message could be simpler, and without emotive content,
* [ ] the message font could be larger,
* [ ] the message could be added to translations,
* [ ] adding sugar3 to puntillism.py is unusual; more common is to keep the file using only pygame,
* [ ] much of the pygame event loop has been duplicated; is there a way to use a function?
* [ ] remove unnecessary print and use logging instead.

srevinsaju commented:
> * [x]  the message could be simpler, and without emotive content,

I personally like emotive content like the messages provided by big source providers which are easier to be understood by children, (and they like it too)
Examples:
![image](https://user-images.githubusercontent.com/48695438/71085525-c6197180-21a8-11ea-9fbf-ac83779a0588.png)
![image](https://user-images.githubusercontent.com/48695438/71085552-d893ab00-21a8-11ea-85eb-e6c55f47a25b.png)
But I fixed it anyway on https://github.com/sugarlabs/pointillism-activity/pull/29/commits/52bdf58a2cf5e53d51d566ea42fa884ddef78b8f

> * [x]  the message font could be larger,

Increased size from 16 to 32.

> * [ ]  the message could be added to translations,

I currently don't know how to use POT files, I assume those are for translations, as I saw en_GB etc

> * [x]  adding sugar3 to puntillism.py is unusual; more common is to keep the file using only pygame,

I added it on mistake, It was not used in code either. I have removed it.

> * [x]  much of the pygame event loop has been duplicated; is there a way to use a function?

After great hardwork :) , I made it into a reusable function so that it can be used many times. This refactoring and formatting on https://github.com/sugarlabs/pointillism-activity/pull/29/commits/4cd13308a1f8dc45cf92ec40817cfca81b19ef77 has paved way to the "Loading Image even if camera present" Feature. Hope you enjoy this one :)

> * [x]  remove unnecessary print and use logging instead.
Removed it.

quozl commented:
> don't know how to use POT files

* wrap the string in the gettext alias, e.g. `_("hello")`
* run the genpot target, e.g. `python setup.py genpot`,
* review and commit the important part of the change to `po/File.pot`, (e.g. with an interactive `git add` or an IDE).

quozl commented:
Please also do flake8 or pep8 on the parts you have changed, and remove whitespace changes that are not required.  Have a look at the [aggregate change](https://github.com/sugarlabs/pointillism-activity/pull/29/files) to see what I mean.

srevinsaju commented:
@quozl, I have fixed all flake8 warning except that of length of line (88), which is too a specific warning.
I did the genpot thingy, I am not sure, if its right. It has created msgid and msgstr in the gettext catalogue.

quozl commented on activity.py:
```diff
@@ -142,21 +151,37 @@ def cradio2_valor(self, radio, value):
+    def save_image(self, image):
         journalobj = datastore.create()
-        journalobj.metadata['title'] = _('Pointillism')
+        journalobj.metadata['title'] = ('Pointillism')
```
Title has lost translation capability.

srevinsaju replied:
https://github.com/sugarlabs/pointillism-activity/pull/29/commits/c74fc8cce206fa3251e8ad50cf5ad0f6be0b7f4b

quozl commented on puntillism.py:
```diff
@@ -47,73 +47,148 @@ def poner_radio2(self, radio):
         self.radio2 = 12
 
+        # create a local variable to check camera not found and in loop
+        self.camNotFound_holder = False
```
New name does not match the style of names used in this file.

srevinsaju replied:
camNotFound_holder is used to keep the loop on hold when camera isn't running. whats a possible alternative? I always can't think beyond `i` and `j` as variable names, but I always fancy on seeing a good name?..

quozl commented on puntillism.py:
```diff
@@ -47,73 +47,148 @@ def poner_radio2(self, radio):
+        blue0 = (0, 0, 128)
+        font = pygame.font.Font('freesansbold.ttf', 32)
+        text = font.render(_('Camera not found'), True, green0, blue0)
+        textRect = text.get_rect()
```
Name not in style.

quozl commented:
Thanks.  Code reviewed.  Some suggestions above as comments.

srevinsaju commented:
@quozl fixed them on https://github.com/sugarlabs/pointillism-activity/pull/29/commits/185eb7a98a19c50a2708d1512ad2a906cf4cfe9e and https://github.com/sugarlabs/pointillism-activity/pull/29/commits/c74fc8cce206fa3251e8ad50cf5ad0f6be0b7f4b

quozl commented:
Thanks.  Tested.  A few things didn't work well for me;

* [ ] if I start the activity, then use F3 to return to the home view, then click on the activity again, the home view becomes a background on which the circles begin to reappear, as if the need to redraw the whole canvas is not detected,
* [ ] pressing the Open Image button stopped the new circles from appearing, but did not pop up an object chooser, and using Alt+Tab or F3 showed the toolbar of Pointillism with other things below it, as if the canvas stops being redrawn when the object chooser is in use,

Other observations;

* [ ] the Camera not found message is green on blue, which would be difficult for colour-blind children to see, it should be white on black,
* [ ] the Open Image icon is different when there is no camera to when there is a camera; and the icon used when there is no camera is "play".

Hope that helps!

srevinsaju commented:
@quozl
>if I start the activity, then use F3 to return to the home view, then click on the activity again, the home view becomes a background on which the circles begin to reappear, as if the need to redraw the whole canvas is not detected,

did you test it with the master brancg.? it's because it's in loop, and call pygame redraw in the while loop, if not, try checking on commit https://github.com/sugarlabs/pointillism-activity/commit/4cd13308a1f8dc45cf92ec40817cfca81b19ef77 , where I didn't port to a function, if that doesn't work either, it's the the error on pygame and other pygame programs.

> pressing the Open Image button stopped the new circles from appearing, but did not pop up an object chooser, and using Alt+Tab or F3 showed the toolbar of Pointillism with other things below it, as if the canvas stops being redrawn when the object chooser is in use,

in the last line of puntillism.py, I stopped the running of the camera loop by making it false on purpose. I can make it run, but I didn't understand the reason why. else to keep it working, I have to go back the the beginning and only show `Camera not Found` and disable the Object Chooser. because, most PyGame Objects do not use Object Chooser. because of the conflict of Pygame and GTK. But let me know if I can do anything else for it. 

 but it's odd the object chooser didn't appear, any logs. i felt there is no need for redrawing while the object chooser is shown, as it will be difficult to stop the loop, if I keep it running at that line.
any logs?

> the Camera not found message is green on blue, which would be difficult for colour-blind children to see, it should be white on black,
 the Open Image icon is different when there is no camera to when there is a camera; and the icon used when there is no camera is "play".

I guess it's not because the camera is not found or found,the code directly uses the open_image icon just like the save_icon is called. I have not created another icon for it. it is directly derived from sugar-artwork, but if you find any alternative icon, I will replace it. you might have tested it on an OLPC and another VBOx or something where the sugar versions are different. the logo for openimage might have changed overtime, causing the difference. if you do have a webcam, try connecting it to the pc you don't have camera, and then launch it to see if it doesn't change.
I too have the Play button, the open thingy, Idk where to get them or use it

thanks for reviewing

quozl commented:
> @quozl
> 
> > if I start the activity, then use F3 to return to the home view, then click on the activity again, the home view becomes a background on which the circles begin to reappear, as if the need to redraw the whole canvas is not detected,
> 
> did you test it with the master brancg.?

I tested with https://github.com/sugarlabs/pointillism-activity/pull/29/commits/c74fc8cce206fa3251e8ad50cf5ad0f6be0b7f4b.

> it's because it's in loop, and call pygame redraw in the while loop, if not, try checking on commit [4cd1330](https://github.com/sugarlabs/pointillism-activity/commit/4cd13308a1f8dc45cf92ec40817cfca81b19ef77) , where I didn't port to a function, if that doesn't work either, it's the the error on pygame and other pygame programs.

Yes, that seems likely.  Can you fix it?

(GTK apps redraw widgets when those widgets are exposed, in response to an exposure event.  PyGame apps usually redraw only when coded to do so.  I guess what needs to happen is an occasional display.update or receive notification that the activity is active.  activity.props.active perhaps.

> > pressing the Open Image button stopped the new circles from appearing, but did not pop up an object chooser, and using Alt+Tab or F3 showed the toolbar of Pointillism with other things below it, as if the canvas stops being redrawn when the object chooser is in use,
> 
> in the last line of puntillism.py, I stopped the running of the camera loop by making it false on purpose. I can make it run, but I didn't understand the reason why. else to keep it working, I have to go back the the beginning and only show `Camera not Found` and disable the Object Chooser. because, most PyGame Objects do not use Object Chooser. because of the conflict of Pygame and GTK. But let me know if I can do anything else for it.

I don't know why the Object Chooser is not visible.  It is occasionally visible when switching between Sugar and the activity.

> but it's odd the object chooser didn't appear, any logs. i felt there is no need for redrawing while the object chooser is shown, as it will be difficult to stop the loop, if I keep it running at that line.
> any logs?

Nothing in logs.

> > the Camera not found message is green on blue, which would be difficult for colour-blind children to see, it should be white on black,
> > the Open Image icon is different when there is no camera to when there is a camera; and the icon used when there is no camera is "play".
> 
> I guess it's not because the camera is not found or found,the code directly uses the open_image icon just like the save_icon is called. I have not created another icon for it. it is directly derived from sugar-artwork, but if you find any alternative icon, I will replace it. you might have tested it on an OLPC and another VBOx or something where the sugar versions are different. the logo for openimage might have changed overtime, causing the difference. if you do have a webcam, try connecting it to the pc you don't have camera, and then launch it to see if it doesn't change.

I can test that next Monday.

> I too have the Play button, the open thingy, Idk where to get them or use it. it's on line 127 for your reference.
> 
> ps: i didn't know sugar supported shortcuts. what about hovering your mouse to the corners and going to home. I have checked that way, the object chooser stays on top however.
> thanks for reviewing

srevinsaju commented:
> I tested with [c74fc8c](https://github.com/sugarlabs/pointillism-activity/commit/c74fc8cce206fa3251e8ad50cf5ad0f6be0b7f4b).

I mentioned the master branch before porting sugarlabs:master = https://github.com/sugarlabs/pointillism-activity

I am checking the rest, will update soon

chimosky commented:
@srevinsaju in `choose_image_from_journal_cb` you should destroy both the `jobject` and `chooser` when you're done with it.

srevinsaju commented:
Ok with commits https://github.com/sugarlabs/pointillism-activity/pull/29/commits/cfd961a3538a798970f675021d0c5750c6a50074, https://github.com/sugarlabs/pointillism-activity/pull/29/commits/89551c8a0cf00d6527f37eed5285405c029fe1ec, and https://github.com/sugarlabs/pointillism-activity/pull/29/commits/474b67dced7a5969cb64d59d0ed3ff9db093df81

I have achieved this:
* Pointillism activity draws over the home screen on switching activities. See @quozl 's comment https://github.com/sugarlabs/pointillism-activity/pull/29#issuecomment-567787000 . This had been existing ever before it was Ported to Python3. Along with this feature, it removes drawing over home screen or any other screen error.
* I tried to close the `jobject` and `ObjectChooser ` as mentioned by @chimosky, first on https://github.com/sugarlabs/pointillism-activity/pull/29/commits/7b05bf8aa32887ad05afffdee2a47ee58e982236 and secondly on https://github.com/sugarlabs/pointillism-activity/pull/29/commits/cfd961a3538a798970f675021d0c5750c6a50074 . On first commit 7b*, I couldn't load image on `pygame.load.image`, instance. (I don't know why). On second commit, I attempted to close the jobject and ObjectChooser , I don't know if it does. 
In the logs, 
* There was an unidentified error, if the file is not chose and the ObjectChooser is simply closed. This might be the error faced by @quozl on https://github.com/sugarlabs/pointillism-activity/pull/29#issuecomment-567787000 . This is finally fixed by giving error message:
(i) If camera is not there, it gives camera not found 
(ii) If camera is there, and if the user presses the load image option, and does not select the image, he gets the message , "Click the Load Image Option"

After this three commits, I have got a personal satisfaction of finishing the task in a better way, than how its supposed to be done. Hope I am right. It boosts my self-confidence.

Here are some animated screenshots for reference. When sugar home is activated with the F3 key, and comes back, it starts redrawing the whole thing , as thats the only possible solution. We can't keep a track of how much of it is completed drawing. And I personally feel, that putting extra effort to use the camera once after the load_image function is called, is not that important, but it can be done. But it requires the user to just restart the activity if he needs to use the camera again. Can be mentioned in the wiki or README.
Hope it works !

![load](https://user-images.githubusercontent.com/48695438/71282554-f101f200-236f-11ea-861d-ede87556a010.gif)
![cam_image](https://user-images.githubusercontent.com/48695438/71282882-08d97600-2370-11ea-8b2c-72db5a535e86.gif)

srevinsaju commented:
@walterbender , @chimosky can you please review this new feature?

srevinsaju commented:
@walterbender , thanks for reviewing. According to the IRC chats, the pointillism `ObjectChooser` remained invisible unless some interaction with the sugar-toolbar or sugar home has taken place. This makes it feel it has been hidden under the `Pointillism-activity`. With commit https://github.com/sugarlabs/pointillism-activity/pull/29/commits/dea253c1541f3bb52aa3142275d6e406d66b7673 the `load_image_from_journal` function was split into two functions, one dedicated for pygame and another for doing the loading process. With this, the journal images load faster than ever. Moreover, with commit https://github.com/sugarlabs/pointillism-activity/pull/29/commits/794e4cc28d4cd02999abb3643d0818796c69d9d7 , it  gives a "Loading Image ..." Text message if the image is being selected.

walterbender commented:
I've not tested your latest commit, but I don't see how it would fix the chooser problem. It seems the chooser is behind the sugargame window. You need to either close sugargame before opening the chooser, or raise it about the sugargame window.

srevinsaju commented:
> I've not tested your latest commit, but I don't see how it would fix the chooser problem. It seems the chooser is behind the sugargame window. You need to either close sugargame before opening the chooser, or raise it about the sugargame window.

It works better on my system, and thats the same way its implemented on colordeducto

walterbender commented:
ASAIK, Color Deducto does not use SugarGame (at least when I first wrote it, it did not).

walterbender commented:
Just tested again. It works!!!

Please run  pep8 and pyflakes on your code. Lots of minor clean up to do.

thx

srevinsaju commented:
Ok thanks @walterbender , glad to know it works! :) Thanks for testing

srevinsaju commented:
@walterbender @quozl I have fixed all the flake8 and pep8 Errors on https://github.com/sugarlabs/pointillism-activity/pull/29/commits/17f8321080b6359f38522a21a6849a9873cc8e66

quozl commented:
Yay!

srevinsaju commented:
:smiley:
